### PR TITLE
Added button and textbox option to SnoreToast with eventHandler

### PR DESCRIPTION
**Pull request by deXol**
Hello,

I added an option for user interaction on toast notifications. If wait option is added, it is processing the user interaction with an eventhandler and prints the choosen button's text or written text to standard output.

### CODE REVIEW

**Comment by tsdgeos** [comment]
Thanks for your contribution :smiley:

This repository is a mirror of a KDE repository. This means that developers are not looking at pull requests created in GitHub, so I'm closing this pull request (actually a bot is doing it).
Please see https://community.kde.org/Infrastructure/Github_Mirror for details on how to contribute to this and other KDE projects.

**Comment by TheOneRing** [comment]
Hi thank you very much. Pls just ignore the bot.

**Comment by tsdgeos** [comment]
Thanks for your contribution :smiley:

This repository is a mirror of a KDE repository. This means that developers are not looking at pull requests created in GitHub, so I'm closing this pull request (actually a bot is doing it).
Please see https://community.kde.org/Infrastructure/Github_Mirror for details on how to contribute to this and other KDE projects.

File: src/main.cpp [hunk 1]
```diff
@@ -150,8 +150,16 @@ SnoreToasts::USER_ACTION parse(std::vector<wchar_t*> args)
         } else  if (arg == L"-tb") {
             isTextBoxEnabled = true;
         } else  if (arg == L"-install") {
-            appID = L"Snore.DesktopToasts";
-            return SUCCEEDED(LinkHelper::tryCreateShortcut(appID)) ? SnoreToasts::Success : SnoreToasts::Failed;
+            std::wstring shortcut(nextArg(it, L"Missing argument to -install.\n"
+                                                     L"Supply argument as -install \"path to your shortcut\" \"path to the application the shortcut should point to\" \"App.ID\""));
+
+            std::wstring exe(nextArg(it, L"Missing argument to -install.\n"
+                                    L"Supply argument as -install \"path to your shortcut\" \"path to the application the shortcut should point to\" \"App.ID\""));
```
Comment (TheOneRing) [inline src/main.cpp]: your missing the help text, could you remove that change in the first place with an interactive rebase?

File: src/linkhelper.cpp [hunk 2]
```diff
@@ -46,11 +48,13 @@ HRESULT LinkHelper::tryCreateShortcut(const std::wstring &shortcutPath, const st
         bool fileExists = attributes < 0xFFFFFFF;
 
         if (!fileExists) {
-            hr = installShortcut(lnkName.str(), exePath, appID);
+            hr = installShortcut(lnkName.str(), exePath, appID, __uuidof(CToastNotificationActivationCallback));
```
Comment (TheOneRing) [inline src/linkhelper.cpp]: I'm not up to date whith the api can you leave a comment here what this does?

File: src/linkhelper.cpp [hunk 3]
```diff
@@ -63,11 +67,25 @@ HRESULT LinkHelper::tryCreateShortcut(const std::wstring &appID)
...
+}
+
 // Install the shortcut
-HRESULT LinkHelper::installShortcut(const std::wstring &shortcutPath, const std::wstring &exePath, const std::wstring &appID)
+HRESULT LinkHelper::installShortcut(const std::wstring &shortcutPath, const std::wstring &exePath, const std::wstring &appID, GUID toastGUID)
 {
+    PCWSTR pszExePath = exePath.c_str();
     std::wcout << L"Installing shortcut: " << shortcutPath << L" " << exePath << L" " << appID << std::endl;
-    HRESULT hr = S_OK;
+    HRESULT hr = HRESULT_FROM_WIN32(::RegSetKeyValueW(HKEY_CURRENT_USER, L"SOFTWARE\\Classes\\CLSID\\{383803B6-AFDA-4220-BFC3-0DBF810106BA}\\LocalServer32", nullptr, REG_SZ, pszExePath, static_cast<DWORD>(wcslen(pszExePath)*sizeof(wchar_t))));
```
Comment (TheOneRing) [inline src/linkhelper.cpp]: so we start a special local server here? the old handles is not sufficient?

File: src/linkhelper.h [hunk 4]
```diff
@@ -18,15 +18,23 @@
 #pragma once
 
 #include "snoretoasts.h"
+#ifndef INIT_PKEY_AppUserModel_ToastActivatorCLSID
+EXTERN_C const PROPERTYKEY DECLSPEC_SELECTANY PKEY_AppUserModel_ToastActivatorCLSID = { { 0x9F4C2855, 0x9F79, 0x4B39,{ 0xA8, 0xD0, 0xE1, 0xD4, 0x2D, 0xE1, 0xD5, 0xF3 } }, 26 };
```
Comment (TheOneRing) [inline src/linkhelper.h]: we can still use custome app ids passed by command line?

File: src/toasteventhandler.cpp [hunk 5]
```diff
@@ -54,40 +59,50 @@ SnoreToasts::USER_ACTION &ToastEventHandler::userAction()
...
 IFACEMETHODIMP ToastEventHandler::Invoke(_In_ IToastNotification * /* sender */, _In_ IToastDismissedEventArgs *e)
 {
+    dismissed = true;
     ToastDismissalReason tdr;
     HRESULT hr = e->get_Reason(&tdr);
     if (SUCCEEDED(hr)) {
         switch (tdr) {
         case ToastDismissalReason_ApplicationHidden:
-            std::wcout << L"The application hid the toast using ToastNotifier.hide()" << std::endl;
+            //std::wcout << L"The application hid the toast using ToastNotifier.hide()" << std::endl;
```
Comment (TheOneRing) [inline src/toasteventhandler.cpp]: why did you comment it out?

**Comment by TheOneRing** [comment]
also 
    SnoreToast.exe -t fpp -m message -w -b "foo;bar"

runs for ever here?

**Comment by TheOneRing** [comment]
when pressing one of the buttons I get
````SnoreToast(9568): onecore\base\appmodel\appxminirepository\lib\appxminirepository.cpp(1161)\kernelbase.dll!00007FFFF6821DD3: (caller: 00007FFFF67CF3C6) ReturnHr(1) tid(808) 80070057 The parameter is incorrect.
SnoreToast(9568): onecore\base\appmodel\appxminirepository\lib\appxminirepository.cpp(1130)\kernelbase.dll!00007FFFF67CF3F8: (caller: 00007FFFF67CEFF0) ReturnHr(2) tid(808) 80070057 The parameter is incorrect.
SnoreToast(9568): onecore\base\appmodel\appxminirepository\lib\appxminirepository.cpp(1009)\kernelbase.dll!00007FFFF6821D60: (caller: 00007FFFF67FDD0B) LogHr(1) tid(808) 80070057 The parameter is incorrect.
SnoreToast(9568): onecore\base\appmodel\identity\lib\packageidentity.cpp(480)\kernelbase.dll!00007FFFF682DE02: (caller: 00007FFFD3C7075D) LogHr(2) tid(808) 80070057 The parameter is incorrect.

File: src/toasteventhandler.cpp [hunk 6]
```diff
@@ -23,8 +23,9 @@
 
 using namespace ABI::Windows::UI::Notifications;
 
-
-static bool resultPrinted = false;
+static bool dismissed = false;
+static bool delaySetEvent = false;
```
Comment (TheOneRing) [inline src/toasteventhandler.cpp]: Whats the point of this commit?

**Comment by deXol** [comment]
Hello,

I will check it next week and cherry pick the necessary commits to a new branch.
The original windows.ui.notifications does not support user interaction from c++ (it has interface for that in c#), so for that it was required to use the CToastNotificationActivationCallback from wrl for buttons and textbox interactions. (print user choices)

**Comment by TheOneRing** [comment]
Awesome thanks! 😊

**Pull request closed without merging**